logic/tool/TypeResolver.java

```diff
@@ -144,355 +84,175 @@ public Conjunction resolveLabels(Conjunction conjunction) {
-                    addNeighbours(newVar, constraint.asPlays().role());
-                    constraint.asPlays().overridden().ifPresent(typeVariable -> addNeighbours(newVar, typeVariable));
-                } else if (constraint.isRelates()) {
-                    newVar.relates(convert(constraint.asRelates().role()),
-                                   constraint.asRelates().overridden().flatMap(var -> Optional.of(convert(var))).orElse(null));
-                    addNeighbours(newVar, constraint.asRelates().role());
-                    constraint.asRelates().overridden().ifPresent(typeVariable -> addNeighbours(newVar, typeVariable));
-                } else if (constraint.isIs()) {
-                    newVar.is(convert(constraint.asIs().variable()));
-                } else throw GraknException.of(ILLEGAL_STATE);
+            Set<ValueType> valueTypes = findComparableValueTypes(constraint);
+            if (valueTypeRegister.get(owner.id()).isEmpty()) {
+                valueTypes.forEach(valueType -> resolverTraversal.valueType(owner.id(), valueType));
+                valueTypeRegister.put(owner.id(), valueTypes);
+            } else if (!valueTypeRegister.get(owner.id()).containsAll(valueTypes)) {
+                throw GraknException.of(SCHEMATICALLY_UNSATISFIABLE_CONJUNCTION, conjunction);
             }
-            return newVar;
+            subAttribute(owner.id());
         }
 
-    }
-
-    private static class Resolvers {
-
-        private final Map<Identifier, TypeVariable> typeResolvers;
-        private final Map<RelationConstraint.RolePlayer, TypeVariable> roleTypeResolvers;
-        // TODO: Why is there no checkers and getters for 'roleTypeResolvers'?
-        // TODO: do we still need 'roleTypeResolvers' given that we no longer put resolveRoleTypes on RelationConstraints?
-
-        private int sysVarCounter;
-
-        Resolvers() {
-            this.sysVarCounter = 0;
-            this.typeResolvers = new HashMap<>();
-            this.roleTypeResolvers = new HashMap<>();
-        }
-
-        TypeVariable register(RelationConstraint.RolePlayer rolePlayer) {
-            return roleTypeResolvers.computeIfAbsent(rolePlayer, rp -> new TypeVariable(newSystemVariable()));
+        private Set<ValueType> findComparableValueTypes(ValueConstraint<?> constraint) {
+            if (constraint.isBoolean()) return set(ValueType.BOOLEAN);
+            else if (constraint.isString()) return set(ValueType.STRING);
+            else if (constraint.isDateTime()) return set(ValueType.DATETIME);
+            else if (constraint.isLong() || constraint.isDouble()) return set(ValueType.LONG, ValueType.DOUBLE);
+            else if (constraint.isVariable()) {
+                TypeVariable comparableVar = convert(constraint.asVariable().value());
+                assert valueTypeRegister.containsKey(comparableVar.id());
+                subAttribute(comparableVar.id());
```

Comment:
At the very least, if it's equal to something it must be some kind of attribute.